cores/esp8266/WString.cpp

```diff
@@ -129,6 +129,7 @@ String::~String() {
 // /*********************************************/
 
 inline void String::init(void) {
+    memset(sso_buf, 0, sizeof(sso_buf));
```

Comment:
I can't see why the initial zeroing in ::init is needed, honestly.  Can you explain?  Once you apply a string of >12 chars to the String, a pointer and 2 length fields will be written over the 0s you wrote in ::init.  At some later point you assign a 1-char string the SSO better write 0 to sso_buff[1] or it'll go poorly...

Reply:
I did test it without this memset and then the garbled texts re-appeared in the use case I showed in the issue.
The main reason I placed this memset in the init, is because that's the only point where I knew for sure that any data present would not be worth keeping.

Immediately after setting this buffer to 0's, the mode is switched to non-SSO.
So then all calls made in init are only on the _ptr struct.
N.B. this struct is then also set to contain 0's, but it does not overlap all the space occupied by `sso_buf`

In the constructor where I had the issues, only the `copy` is called after `init`.
This calls `reserve`, which calls `changeBuffer`.

In `changeBuffer` we enter this part, since the `ptr.buff` is set to 0 in `init`:

https://github.com/esp8266/Arduino/blob/db88c828824f38db55e863274137d0f599011511/cores/esp8266/WString.cpp#L156-L162

So only SSO is set and nothing else.
Since nothing has changed to the SSO buffer, but SSO is now active, a call to `len()` may return something other than 0 (it is calling `strlen(sso_buf)` which may return anything)

So the first byte of the buffer is not set to 0 and thus the String object may contain garbage and it is up to the compiler how this union is aligned.
If the pointer of `_ptr` is at the first byte of the `sso_buf`, then it should already be a zero, but apparently the Windows build is not locating the `_prt` struct at the same address as `sso_buf`.

By wiping as much as possible to 0 in this union struct, it is guaranteed it will contain zeroes at the start.
Also I guess it could differ between builds where this union is located and I don't know what all possible offsets are between the start of `_ptr` and `sso_buf`.

So that's why I think it is needed to have the `sso_buf` initialized to 0.
Does that make any sense?

Reply:
AMD64 compiles are going to align things somewhat differently than X86_32 ones do (the pointer, for example, is gonna be 2x the size...).  I don't know if Windows supports it w/the mingw(?) toolchain, but there is a 32-bit compile mode in the makefile which should help get closer to on-board alignment.

I'm not sure I fully follow, but my point is in the general case the sequence you've got is not guaranteed to happen, and there's possibility for the early bytes of the union to contain non-0s after you've called ::init and when switching.  0-ing out at ::init is safe, but I think it's unneeded.

I think the main problem is you found that when a SSO string was copied into the struct it did not get explicitly terminated w/a \0.  Your patch does fix that, later on in the SSO assignment changeBuffer.  My gut says that dropping this memset but leaving everything else untouched won't change the behavior.

Reply:
Well, leaving out this memset did not fix the issue (tested it).
I'm also not 100% convinced that whatever this init does fix will cover all situations, but at least it does something.
Maybe the suggestion I made yesterday does cover all and then this part in the init can be removed?

I don't understand what you tried to explain with the difference on 64-bit and 32 bit platforms.
Do the compilers generate different code even though they compile for the same 32 bit platform?
I know the `size_t`, `ptrdiff_t` and pointers have different sizes when compiling for 64 bit or 32 bit platforms, but when they compile for the same 32 bit platform they should align structures in memory the same, right?
Otherwise we might have a problem when serializing structs.

Reply:
I was talking about the host testing.  It can run either 32-bit or 64-bit on your system, depending on the compilers available and `make` options. The `gcc-xtensa` binaries should produce the exact same code under either host OS, of course.